### PR TITLE
GH-35871: [Go] Account for struct validity bitmap in `array.ApproxEqual`

### LINKED ISSUE #35871: `array.ApproxEqual` should take parent struct validity bitmap prior to checking fields
### Describe the enhancement requested

When comparing `array.Struct` values with `array.ApproxEqual` the validity bitmap of the struct itself should take precedence:
> When reading the struct array the parent validity bitmap takes priority.

### Component(s)

Go

### PR DESCRIPTION
**Pull request by candiduslynx**
### Rationale for this change

When comparing `array.Struct` values with `array.ApproxEqual` the validity bitmap of the struct itself should take precedence:
> When reading the struct array the parent validity bitmap takes priority.

This follows a brief discussion from #35851.

### What changes are included in this PR?

`array.arrayApproxEqualStruct` will check the fields data only if the struct elem is valid.

### Are these changes tested?

`pqarrow` tests are updated accordingly (no special treatment for structs, just `array.ApproxEqual`

### Are there any user-facing changes?

`array.ApproxEqual` behavior changed to match the docs about validity bitmap precedence.

* Closes: #35871

### CODE REVIEW

**Comment by candiduslynx** [comment]
@zeroshade this PR includes some changes to pqarrow pkg as well, but they are minor (exit early on the found value, dedup code).
I can extract it into a minor PR, if that's required.

File: go/arrow/array/compare.go [hunk 1]
```diff
@@ -735,10 +735,14 @@ func arrayApproxEqualFixedSizeList(left, right *FixedSizeList, opt equalOption)
...
 
 func arrayApproxEqualStruct(left, right *Struct, opt equalOption) bool {
-	for i, lf := range left.fields {
-		rf := right.fields[i]
-		if !arrayApproxEqual(lf, rf, opt) {
-			return false
+	for i := 0; i < left.Len(); i++ {
+		if left.IsNull(i) {
+			continue
+		}
```
Comment (zeroshade) [inline go/arrow/array/compare.go:741]: shouldn't this be:

```go
if left.IsNull(i) {
    if !right.IsNull(i) {
        return false
    }
    continue
}
```
?

**Review by zeroshade — CHANGES_REQUESTED** [review]

File: go/arrow/array/compare.go (hunk 1 above)
Comment (candiduslynx) [reply go/arrow/array/compare.go:741]: we already checked for this in `baseArrayEqual`, see other `approx` implementations

File: go/arrow/array/compare.go [hunk 2]
```diff
@@ -735,10 +735,14 @@ func arrayApproxEqualFixedSizeList(left, right *FixedSizeList, opt equalOption)
...
+		for j := range left.fields {
+			if !sliceApproxEqual(left.fields[j], int64(i), int64(i+1), right.fields[j], int64(i), int64(i+1), opt) {
+				return false
+			}
 		}
```
[the remark this answers is not in the document]
Comment (candiduslynx) [reply go/arrow/array/compare.go]: Done in fb156f8a1198495134b0841a2c4021f6b1c6f619

**Comment by candiduslynx** [comment]
@zeroshade I've updated the code to use bitutils visit, thanks for the suggestion!

**Review by zeroshade — APPROVED** [review]
LGTM thanks!

**Pull request merged**